optuna/storages/grpc/_server.py

```diff
@@ -0,0 +1,486 @@
+        state=_to_proto_trial_state(frozen_trial.state),
+        values=frozen_trial.values,
+        datetime_start=(
+            frozen_trial.datetime_start.strftime(DATETIME_FORMAT)
+            if frozen_trial.datetime_start
+            else ""
+        ),
+        datetime_complete=(
+            frozen_trial.datetime_complete.strftime(DATETIME_FORMAT)
+            if frozen_trial.datetime_complete
+            else ""
+        ),
+        distributions={
+            key: distribution_to_json(distribution)
+            for key, distribution in frozen_trial.distributions.items()
+        },
+        params=params,
+        user_attributes={key: json.dumps(value) for key, value in frozen_trial.user_attrs.items()},
+        system_attributes={
+            key: json.dumps(value) for key, value in frozen_trial.system_attrs.items()
+        },
+        intermediate_values={
+            step: value for step, value in frozen_trial.intermediate_values.items()
+        },
+    )
+
+
+def _from_proto_frozen_trial(frozen_trial: api_pb2.FrozenTrial) -> FrozenTrial:
+    datetime_start = (
+        datetime.strptime(frozen_trial.datetime_start, DATETIME_FORMAT)
+        if frozen_trial.datetime_start
+        else None
+    )
+    datetime_complete = (
+        datetime.strptime(frozen_trial.datetime_complete, DATETIME_FORMAT)
+        if frozen_trial.datetime_complete
+        else None
+    )
+    distributions = {
+        key: json_to_distribution(value) for key, value in frozen_trial.distributions.items()
+    }
+    params = {}
+    for key, value in frozen_trial.params.items():
+        params[key] = distributions[key].to_external_repr(value)
+
+    return FrozenTrial(
+        trial_id=frozen_trial.trial_id,
+        number=frozen_trial.number,
+        state=_from_proto_trial_state(frozen_trial.state),
+        value=None,
+        values=frozen_trial.values if frozen_trial.values else None,
+        datetime_start=datetime_start,
+        datetime_complete=datetime_complete,
+        params=params,
+        distributions=distributions,
+        user_attrs={key: json.loads(value) for key, value in frozen_trial.user_attributes.items()},
+        system_attrs={
+            key: json.loads(value) for key, value in frozen_trial.system_attributes.items()
+        },
+        intermediate_values={
+            step: value for step, value in frozen_trial.intermediate_values.items()
+        },
+    )
+
+
+def make_server(
+    storage: BaseStorage, host: str, port: int, thread_pool: ThreadPoolExecutor | None = None
+) -> grpc.Server:
+    server = grpc.server(thread_pool or ThreadPoolExecutor(max_workers=10))
+    api_pb2_grpc.add_StorageServiceServicer_to_server(
+        OptunaStorageProxyService(storage), server
+    )  # type: ignore
+    server.add_insecure_port(f"{host}:{port}")
+    return server
+
+
+def run_grpc_server(
+    storage: BaseStorage, host: str, port: int, thread_pool: ThreadPoolExecutor | None = None
+) -> None:
```

Comment:
I have two nit-picky suggestions.

1. Consider making the `host` and `port` arguments optional by providing default values like the constructor method for `GrpcStorageProxy`.
2. I prefer `run_grpc_proxy_server` over `run_grpc_server`, as the latter feels too generic.

```suggestion
def run_grpc_proxy_server(
    storage: BaseStorage,
    *,
    host: str = "127.0.0.1",
    port: int = 13000,
    thread_pool: ThreadPoolExecutor | None = None
) -> None:
```